### PR TITLE
[Web] Processing refactor - DOM element interfaces

### PR DESCRIPTION
**Pull request by jahorton**
This PR will serve to provide the core KeymanWeb processing engine a clean abstraction for text processing interactions with all supported DOM elements, additionally adding support for a new `Mock` output type useful for some of our future goals.  It is based on #1558, which refactors touch alias elements - originally that was included here, but the changes were large and numerous enough to merit spinning that work off from this PR.  As this PR defines types dependent on the newly-refactored target, the resulting PR chain is necessary.

The core abstraction is defined in the interface `EditableElement` - when properly implemented, both `KC_` (the core context function) and `output` (the core keyboard output function) will be rewritten to target this interface's methods.  To facilitate this, some of the interface's implementers will need minor support from DOM events for state maintenance, but nothing worse than what KMW already does.  These ties will be estabilshed toward the end of this PR's development.

An extra benefit - establishing unit testing on KMW's output to the DOM!  This interface and its implementers can be compiled separately from KMW.  By validating them through unit testing, we can ensure that the abstraction will work flawlessly and more easily address DOM-interaction bugs that affect specific supported target element types.  This will be used to validate each of the interface's implementing classes before tying them properly into KMW itself.

This will also lead us one step closer to 'headless' engine testing, as implementing a `Mock` output will facilitate allowing keyboards to operate outside of the browser/DOM, though that target is for another PR further down the line.

During work on this, it was also discovered that #504 failed to properly handle design-mode iframe command caching; this has now been corrected.

Not implemented here:
* the `Mock` / headless element interface type.
  - See #1571.
* Deadkeys should probably be included as state fields of each element interface and cleared upon change of focused element.
  - This will be very useful for 'snapshotting' the context's state for use in predictive text and autocorrective scenarios.
  - See #1569.

Issues:
* Design-mode iframes work, but do not display the KMW UI nor the OSK.  (If a keyboard is selected before focusing the iframe, that keyboard may be used with physical keystrokes.)
  - This is likely a pre-existing issue, though I have yet to confirm this with a test against the beta branch.

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/139108:id/kvk/index.html)

File: web/source/kmwcallback.ts [hunk 1]
```diff
@@ -915,8 +920,8 @@ namespace com.keyman {
           var _CacheableCommands = this._CacheCommands(Ldoc);
         }
       
-        var Lsel = Ldv.getSelection();
-        var LselectionStart = Lsel.focusNode.nodeValue ? Lsel.focusNode.substringData(0,Lsel.focusOffset)._kmwLength() : 0;  // I3319
+        var Lsel: Selection = Ldv.getSelection();
+        var LselectionStart = Lsel.focusNode.nodeValue ? (<Text> Lsel.focusNode).substringData(0,Lsel.focusOffset)._kmwLength() : 0;  // I3319
```
Comment (jahorton) [inline web/source/kmwcallback.ts]: Current edits in this file exist to aid analysis of existing code, which should facilitate before-and-after comparisons with the new classes.

File: web/unit_tests/cases/element_interfaces.js [hunk 2]
```diff
@@ -150,25 +150,37 @@ if(typeof InterfaceTests == 'undefined') {
...
+          setIESelection(node, sel, start, end);
         } else {
+          sel.removeAllRanges();
           // Does not work on IE!
-          sel.setPosition(node, start);
-          sel.extend(node, end);
+          try {
+            sel.setPosition(node, start);
+            sel.extend(node, end);
+          } catch (e) {
```
Comment (jahorton) [inline web/unit_tests/cases/element_interfaces.js]: Addresses a weird testing-setup error that's affecting Firefox when run through BrowserStack.  It's only the one browser, so at least the affected cases (reverse-selection correctness) are well-covered by Chrome, Edge, and Safari.

File: web/source/build_dev_resources.sh [hunk 3]
```diff
@@ -2,9 +2,6 @@
 # 
 # Compiles development-related KeymanWeb resources for use with developing/running tests.
 #
-# Includes:
-#   - the Recorder module (for engine tests)
-#   - the DOM module (for touch-alias tests)
```
Comment (jahorton) [inline web/source/build_dev_resources.sh]: TODO:  reinstate this.

File: web/source/kmwcallback.ts [hunk 4]
```diff
@@ -874,139 +876,30 @@ namespace com.keyman {
...
-          LselectionStart = Pelem._KeymanWebSelectionStart;
-          LselectionEnd = Pelem._KeymanWebSelectionEnd;
-        } else {
-          LselectionStart = Pelem.value._kmwCodeUnitToCodePoint(Pelem.selectionStart);  // I3319
-          LselectionEnd = Pelem.value._kmwCodeUnitToCodePoint(Pelem.selectionEnd);      // I3319
-        }
-        
-        var LscrollTop, LscrollLeft;
-        if(Pelem.type.toLowerCase() == 'textarea' && typeof(Pelem.scrollTop) != 'undefined') {
-          LscrollTop = Pelem.scrollTop; LscrollLeft = Pelem.scrollLeft;
```
Comment (jahorton) [inline web/source/kmwcallback.ts]: TODO:  Reinstate `TextArea` scroll caching!

**Pull request merged**